add episode on contact matrices (#63)
epiverse-trace/tutorials-late: episodes/contact-matrices.Rmd

### In mathematical models

Consider the SIR model where individuals are categorized as either susceptible $S$, infected but not yet infectious $E$, infectious $I$ or recovered $R$. The schematic below shows the processes which describe the flow of individuals between the disease states $S$, $I$ and $R$ and the key parameters for each process.

```{r diagram, echo = FALSE, message = FALSE}
DiagrammeR::grViz("digraph {

  # graph statement
  #################
  graph [layout = dot,
         rankdir = LR,
         overlap = true,
         fontsize = 10]

  # nodes
  #######
  node [shape = square,
       fixedsize = true
       width = 1.3]

       S
       I
       R

  # edges
  #######
  S -> I [label = ' infection \n(transmission rate &beta;)']
  I -> R [label = ' recovery \n(recovery rate &gamma;)']

}")
```

The [differential equations](../learners/reference.md#ordinary) below describe how individuals move from one state to another [(Bjørnstad et al. 2020)](https://doi.org/10.1038/s41592-020-0822-z).

$$
\begin{aligned}

> \frac{dS}{dt} & = - \beta S I \\

Reviewer: Thanks, so there's a $1/N$ in the $\beta$ term? I think sentences like these are still consistent then (as implicitly doing $I/N$ but with the $1/N$ wrapped into $\beta$): `we want to calculate the proportion of individuals infected in each group $j$, multiplied by the number of contacts made with group $i$, multiplied by the transmission risk per contact`

> \frac{dS}{dt} & = - \beta S I \\

Author: Thanks for highlighting this, it took me down a bit of a rabbit hole which was useful! I wrote these equations based on the [model code](https://github.com/epiverse-trace/epidemics/blob/f706261bac19282e813ca43675ff16dcb79b3d40/R/model_default.R#L346) which has density dependent transmission - which as you say incorporates (1/N) into the transmission term. Thinking about this, it has implications for the calculation of the transmission rate from R0, so there needs to be 1/N included somewhere. In the contact matrix preparation there is a division by [population size](https://github.com/epiverse-trace/epidemics/blob/f706261bac19282e813ca43675ff16dcb79b3d40/R/tools.R#L77). Is the purpose of this step to account for 1/N in the transmission term?

> \frac{dS}{dt} & = - \beta S I \\

Reviewer: Yes, I think that division by N is for the transmission term, but worth explaining this as I've wondered in the past where the different assumptions are feeding into the pre-processing steps!

\frac{dI_i}{dt} &= \beta S_i\sum_j C_{j,i} I_j - \gamma I_i \\
\frac{dR_i}{dt} &=\gamma I_i \\
\end{aligned}
$$

::::::::::::::::::::::::::::::::::::: callout
##  Normalising the contact matrix to ensure the correct value of $R_0$

When simulating an epidemic, we often want to ensure that the average number of secondary cases generated by a typical infectious individual (i.e. $R_0$) is consistent with known values for the pathogen we're analysing. In the above model, we scale the contact matrix by the $\beta$ to convert the raw interaction data into a transmission rate. But how do we define the value of $\beta$ to ensure a certain value of $R_0$?

Rather than just using the raw number of contacts, we can instead normalise the contact matrix to make it easier to work in terms of $R_0$. Normalisation means converting to a value to be between 0 and 1. In particular, we normalise the matrix by scaling it so that if we were to calculate the average number of secondary cases based on this normalised matrix, the result would be 1 (in mathematical terms, we are scaling the matrix so the largest eigenvalue is 1). This transformation scales the entries but preserves their relative values.

> In the case of the above model, we want to define `beta * C[j,i]` so that the model has a specified valued of $R_0$. We do this by first transposing the matrix (so row/column entries are now `C[j,i])`, then normalise it so the maximum eigenvalue is one. Because the rate of recovery is $\gamma$, individuals will be infectious on average for $1/\gamma$ days. So $\beta$ is calculated from the scaling factor and the value of $\gamma$  (i.e. mathematically we have the dominant eigenvalue of the matrix $R_0 \times C_{normalised}$ is $\beta / \gamma$).

Author: I have made some edits to distinguish C from `contact_data$matrix` as suggested - I think it's consistent now

::::::::::::::::::::::::::::::::::::::::::::::::

The example above uses the POLYMOD survey. There are a number of surveys available in `socialmixr`, to list the available surveys use `list_surveys()`. To download a survey, we can use `get_survey()`

```{r, message = FALSE, warning = FALSE}
zambia_sa_survey <- get_survey("https://doi.org/10.5281/zenodo.3874675")
```

::::::::::::::::::::::::::::::::::::: challenge 

## Zambia contact matrix

After downloading the survey, generate a symmetric contact matrix for Zambia using the following age bins:

> + [0,15)

Reviewer 2: ```suggestion
+ [0,20)
```

+ and in mathematical models of tranmission within a population, to account for group specific contact patterns.

However, all of these applications require us to perform some additional calculations using the contact matrix. Specifically, there are two main calculations we often need to do:

1. **Convert contact matrix into expected number of secondary cases**

If contacts vary between groups, then the average number of secondary cases won't be equal simply to the average number of contacts multiplied by the probability of transmission-per-contact. This is because the average amount of transmission in each generation of infection isn't just a matter of whom a group came into contact with; it's about whom *their contacts* subsequently come into contact with. The function `r_eff` in the package `{finalsize}` can perform this conversion, taking a contact matrix, demography and proportion susceptible and converting it into an estimate of the average number of secondary cases generated by a typical infectious individual (i.e. the effective reproduction number).

2. **Convert contact matrix into contact rates**

> Whereas a contact matrix gives the average number of contacts that one groups makes with another, epidemic dynamics in different groups depend on the rate at which one group infects another. We therefore need to scale the rate of interaction between different groups (i.e. the number of contacts per unit time) to get the rate of transmission. However, we need to be careful that we are defining transmission to and from each group correctly in any model. Specifically, the entry $(i,j)$ in a mathematical model contact matrix represents contacts by group $i$ made with group $j$. But if we want to know the rate at which a group $i$ are getting infected, then we want to calculate the proportion of individuals infected in each group $j$ ($I_j/N_j$), multiplied by the number of contacts made with group $i$ ($C[i,j]$), multiplied by the transmission risk per contact ($\beta$).

Reviewer 2: Something is flipped between
- "the entry $(i,j)$ in a mathematical model contact matrix represents contacts by group $i$ made with group $j$"
- "the number of contacts made with group $i$ ($C[i,j]$)"
In an attempt to phrase this clearer (acknowledging how hard it is to communicate this):

But if we want to know the rate at which a group $i$ are getting infected, then we want to calculate the number of contacts in group $j$ made by susceptibles in group $i$ ($S_i C[i,j]$) with the proportion of those contacts that are infectious ($I_j/N_j$), and the transmission risk per contact ($\beta$).

Reviewer 2: You probably noticed this but re-reading it this should probably be "calculate" -> "multiply"

$$ \beta S_i \sum_j C_{i,j} I_j/N_j. $$ 

Susceptible individuals in age group $i$ become infected dependent on their rate of contact with individuals in each age group. For each disease state ($S$, $E$, $I$ and $R$) and age group ($i$), we have a differential equation describing the rate of change with respect to time.  

$$
\begin{aligned}
\frac{dS_i}{dt} & = - \beta S_i \sum_j C_{i,j} I_j/N_j \\
\frac{dI_i}{dt} &= \beta S_i\sum_j C_{i,j} I_j/N_j - \gamma I_i \\
\frac{dR_i}{dt} &=\gamma I_i \\
\end{aligned}
$$

::::::::::::::::::::::::::::::::::::: callout
##  Normalising the contact matrix to ensure the correct value of $R_0$

When simulating an epidemic, we often want to ensure that the average number of secondary cases generated by a typical infectious individual (i.e. $R_0$) is consistent with known values for the pathogen we're analysing. In the above model, we scale the contact matrix by the $\beta$ to convert the raw interaction data into a transmission rate. But how do we define the value of $\beta$ to ensure a certain value of $R_0$?

> Rather than just using the raw number of contacts, we can instead normalise the contact matrix to make it easier to work in terms of $R_0$. Normalisation means converting to a value to be between 0 and 1. In particular, we normalise the matrix by scaling it so that if we were to calculate the average number of secondary cases based on this normalised matrix, the result would be 1 (in mathematical terms, we are scaling the matrix so the largest eigenvalue is 1). This transformation scales the entries but preserves their relative values.

Reviewer 2: "Normalisation means converting to a value to be between 0 and 1." - not sure we need this statement? As you explain in more detail in the following sentence what is happening

> In the case of the above model, we want to define $\beta  C_{i,j}$ so that the model has a specified valued of $R_0$. $C[i,j]$ is defined as contacts to $i$ from $j$, which is equivalent to `contact_data$matrix[j,i]` so the first step is to transpose the contact data matrix (`contact_data$matrix`) so the row/column entries are now in the form $C[i,j]$. Then we normalise the matrix $C$ so the maximum eigenvalue is one and call this matrix $C_normalised$. Because the rate of recovery is $\gamma$, individuals will be infectious on average for $1/\gamma$ days. So $\beta$ is calculated from the scaling factor and the value of $\gamma$  (i.e. mathematically we have the dominant eigenvalue of the matrix $R_0 \times C_{normalised}$ is $\beta / \gamma$).

Reviewer 2: - "defined as contacts to $i$ from $j$" - I don't think this from/to notion is useful as it implies that it's one of the two that initiates contact - see also discussion at https://github.com/socialcontactdata/contactmatrix/issues/14
- you could (if you don't think it adds confusion) point to the `split` argument in `contact_matrix` which does the normalisation for you, although it's definitely also useful to show here how to do iot.

Author: I've altered the text slightly to 'represents the contacts between populations $i$ and $j$' so it is more generic, wording this episode has been tricky - if you have any more suggestions for clarity let me know!

I've also added a callout on using `split`, I think it is a useful addition to know the normalisation can be done within the function `contact_data()`. Related to this, I think there could be come confusion about where normalisation takes place in different analyses e.g. in epidemics it happens [within the model function call ](https://github.com/epiverse-trace/epidemics/blob/f706261bac19282e813ca43675ff16dcb79b3d40/R/tools.R#L77), I've added a callout box to the first tutorial on using `epidemics` to highlight that the contact matrix normalisation does not need to be done.

Reviewer 2: Re-reading the various uses here I quite like your earlier use of contacts _of_ group i _with_ group j which to me does not imply any directionality but makes it clear that in rows we're specifically looking at group i. So perhaps adopt this one throughout?

+ and in mathematical models of transmission within a population, to account for group specific contact patterns.

However, all of these applications require us to perform some additional calculations using the contact matrix. Specifically, there are two main calculations we often need to do:

1. **Convert contact matrix into expected number of secondary cases**

If contacts vary between groups, then the average number of secondary cases won't be equal simply to the average number of contacts multiplied by the probability of transmission-per-contact. This is because the average amount of transmission in each generation of infection isn't just a matter of whom a group came into contact with; it's about whom *their contacts* subsequently come into contact with. The function `r_eff` in the package `{finalsize}` can perform this conversion, taking a contact matrix, demography and proportion susceptible and converting it into an estimate of the average number of secondary cases generated by a typical infectious individual (i.e. the effective reproduction number).

2. **Convert contact matrix into contact rates**

> Whereas a contact matrix gives the average number of contacts that one groups makes with another, epidemic dynamics in different groups depend on the rate at which one group infects another. We therefore need to scale the rate of interaction between different groups (i.e. the number of contacts per unit time) to get the rate of transmission. However, we need to be careful that we are defining transmission to and from each group correctly in any model. Specifically, the entry $(i,j)$ in a mathematical model contact matrix represents contacts by group $i$ made with group $j$. But if we want to know the rate at which a group $i$ are getting infected, then we want to multiply the number of contacts in group $j$ ($C[i,j]$) made by susceptibles in group $i$ ($S_i$) with the proportion of those contacts that are infectious ($I_j/N_j$), and the transmission risk per contact ($\beta$).

Reviewer 2: this maybe?
```suggestion
Whereas a contact matrix gives the average number of contacts that one groups makes with another, epidemic dynamics in different groups depend on the rate at which one group infects another. We therefore need to scale the rate of interaction between different groups (i.e. the number of contacts per unit time) to get the rate of transmission. However, we need to be careful that we are defining transmission to and from each group correctly in any model. Specifically, the entry $(i,j)$ in a mathematical model contact matrix represents contacts of group $i$ with group $j$. But if we want to know the rate at which a group $i$ are getting infected, then we want to multiply the number of contacts of susceptibles in group $i$ ($S_i$) with group $j$ ($C[i,j]$) with the proportion of those contacts that are infectious ($I_j/N_j$), and the transmission risk per contact ($\beta$).
```

$$
\begin{aligned}
\frac{dS_i}{dt} & = - \beta S_i \sum_j C_{i,j} I_j/N_j \\
\frac{dI_i}{dt} &= \beta S_i\sum_j C_{i,j} I_j/N_j - \gamma I_i \\
\frac{dR_i}{dt} &=\gamma I_i \\
\end{aligned}
$$

###  Normalising the contact matrix to ensure the correct value of $R_0$

When simulating an epidemic, we often want to ensure that the average number of secondary cases generated by a typical infectious individual (i.e. $R_0$) is consistent with known values for the pathogen we're analysing. In the above model, we scale the contact matrix by the $\beta$ to convert the raw interaction data into a transmission rate. But how do we define the value of $\beta$ to ensure a certain value of $R_0$?

Rather than just using the raw number of contacts, we can instead normalise the contact matrix to make it easier to work in terms of $R_0$. In particular, we normalise the matrix by scaling it so that if we were to calculate the average number of secondary cases based on this normalised matrix, the result would be 1 (in mathematical terms, we are scaling the matrix so the largest eigenvalue is 1). This transformation scales the entries but preserves their relative values.

> In the case of the above model, we want to define $\beta  C_{i,j}$ so that the model has a specified valued of $R_0$. The entry of the contact matrix $C[i,j]$ represents the contacts between populations $i$ and $j$, which is equivalent to `contact_data$matrix[j,i]` so the first step is to transpose the contact data matrix (`contact_data$matrix`) so the row/column entries are now in the form $C[i,j]$. Then we normalise the matrix $C$ so the maximum eigenvalue is one and call this matrix $C_normalised$. Because the rate of recovery is $\gamma$, individuals will be infectious on average for $1/\gamma$ days. So $\beta$ is calculated from the scaling factor and the value of $\gamma$  (i.e. mathematically we have the dominant eigenvalue of the matrix $R_0 \times C_{normalised}$ is $\beta / \gamma$).

Reviewer 2: ```suggestion
In the case of the above model, we want to define $\beta  C_{i,j}$ so that the model has a specified valued of $R_0$. The entry of the contact matrix $C[i,j]$ represents the contacts of population $j$ with population $i$, which is equivalent to `contact_data$matrix[j,i]` so the first step is to transpose the contact data matrix (`contact_data$matrix`) so the row/column entries are now in the form $C[i,j]$. Then we normalise the matrix $C$ so the maximum eigenvalue is one and call this matrix $C_normalised$. Because the rate of recovery is $\gamma$, individuals will be infectious on average for $1/\gamma$ days. So $\beta$ is calculated from the scaling factor and the value of $\gamma$  (i.e. mathematically we have the dominant eigenvalue of the matrix $R_0 \times C_{normalised}$ is $\beta / \gamma$).
```

Reviewer 2: PS: I think this is the wrong way around - C[i, j] in socialmixr is as in the equations above I think (but please correct me if I'm wrong).

Author: If it's the same then why do we need to transpose the matrix?

Reviewer: Always find $i$ and $j$ a potential headache (which it's why this will be so useful to have written down!)

Taking step back, we just need the FOI to be defined sensibly, i.e.: $\sum_j C_{i,j} I_j/N_j$. So $C_{i,j}$ should be contacts that group $j$ (the infectious ones) make with group $i$ (the susceptible ones) - this is from equation A3 in [Wallinga et al (2006)](https://pubmed.ncbi.nlm.nih.gov/16968863/)

The `contact_matrix()` function gives the following structure:
```
#> $matrix
#>          contact.age.group
#> age.group      [0,1)     [1,5)   [5,15)      15+
#>    [0,1)  0.40000000 0.8000000 1.266667 5.933333
#>    [1,5)  0.11250000 1.9375000 1.462500 5.450000
#>    [5,15) 0.02450980 0.5049020 7.946078 6.215686
#>    15+    0.03230337 0.3581461 1.290730 9.594101
```
So $C[i,j]$ is contacts made by group $i$ with group $j$? Which I think would mean it needs transposing?

For completeness (and just to remind myself), {epidemics} has this internal processing in `.prepare_population()`, which normalises by dominant eigenvalue and scales based on $w_{tot}/w_i$ (to use the Walling et al notation):
```
contact_matrix <- (contact_matrix / max(Re(eigen(contact_matrix)$values))) /
    x[["demography_vector"]]
```

Reviewer 2: > we just need the FOI to be defined sensibly, i.e.: $\sum_j C_{i,j} I_j/N_j$. So $C_{i,j}$ should be contacts that group $j$ (the infectious ones) make with group $i$ (the susceptible ones) 

Shouldn't this be the other way round, i.e. $C_{ij}$ here is the average number of contacts with group $j$ that a suspectible in group $i$ has (then multiplied with the probability that the contact is infectious, i.e. $I_j/N_j$)?

Here's an example:

There are two groups, $i and $j$. There are 1 person in group $i$ ($N_i=1$) and 100 people in group $j$ ($N_j = 100$). Person $i$ meets all 100 people in group $j$ every day: according to socialmixr notation that means $C_{ij}=100$ and $C_{ji}=1$ (on the scale of days). The total number of contacts between the two groups per day is $C_{ij} N_i = C_{ji} N_j = 100$.

Sticking with this notation:
The FOI on person $i$ is proportional to 100 * (proportion of $j$ that is ill), or $C_{ij} I_j / N_j$
The FOI on people in group $j$ is proportional to 1 * (1 if $i$ is ill, 0 otherwise), or $C_{ji} I_i / N_i$

Reviewer 2: > Always find $i$ and $j$ a potential headache (which it's why this will be so useful to have written down!)

Can definitely agree on that!

Reviewer: This is a useful illustration, thanks. That does make more sense given it's FOI onto the susceptible (and maybe we should include as an example once converge on answer!) Suspect I was thinking in terms of 'outbound' contacts (as in the next generation matrix and R estimation case) rather than 'inbound' contacts for the SIR model.

I feel I'm still missing something with regards to the Wallinga framing though, because those are the contacts made per person by group $j$ with group $i$ (i.e. $m_{ij}$ to use the Wallinga et al notation), rather than per capita contact rate $c_{ij} = m_ij / w_i$ where $w_i$ is the proportion of population in group $i$, and equal to $\beta_{ij}$ but for a scaling factor. So in above example, the contact rate $c_{ij} = 1 \times 101/1 \approx 100$ and $c_{ji} = 100 \times 101/100 \approx 100$.

Reviewer 2: Sticking with $C_{ij} = m_{ji}$ and $N_i=\omega_i$ and $I_i=y_i$ (LHS: socialmixr, RHS: Wallinga) we have for the term in the sum above:

$$
C_{ij} \frac{I_j}{N_j} = m_{ji} \frac{y_j}{\omega_j} = m_{ij} \frac{\omega_j}{\omega_i} \frac{y_j}{\omega_j} = m_{ij} \frac{y_j}{\omega_i} 
$$

Where we've used the symmetry $m_{ji} \omega_i = m_{ij} \omega_j$. That's the same as the equation under "Log-likelihood functions for the transmission parameters" in the Wallinga et al. paper (noting that we've swapped the index in the denominator).

Reviewer: Thanks, that makes sense. So it basically comes down to whether the $N_j$ (or symmetry-derived equivalent) is wrapped into the $\beta_{ij}$ term. If defined separately, i.e. $\sum_j C_{i,j} I_j/N_j$ as above, then as you say, contact rate should be defined from-S-to-I.

Reviewer 2: Not sure if this is useful for the purpose of teaching but I find it easier sometimes to work from the symmetric encounter matrix, i.e. the number of encounters between group $i$ and group $j$ per unit time. If we call this $E_{i,j}$ then it is symmetric $E_{i,j}= E_{j,i}$ and so is the term in the force of infection which is proportional to $\frac{E_{ij}I_j}{N_iN_j}$. This highlights that the row vs. column notation is purely about how the matrix is normalised i.e. whether we write it as $\frac{E_{ij}}{N_i}$ or $\frac{E_{ij}}{N_j}$ (which thus determines which of the $N$ terms remains in the force of infection) and not about contacts from/to etc.

Reviewer: I've gone through the {epidemics} and {finalsize} examples a bit more, thinking about how these steps are introduced in vignettes. The challenge is that these packages allow user to specify in terms of $R_0$ (which is useful), but that means normalising the matrix, then converting back into the correct form for the contact rate you describe above @sbfnk . This is where I think the transpose comes in, so that it's switching between $C_{ij}=R_{ij}$ for the eigenvalue normalisation and $C_{ij}/N_j$ (i.e. contact rate per capita) for the model.

But because the result is the symmetric per capita matrix that goes into the model, the end result is equivalent:
```r
# 1. Current epidemics and finalsize implementation
contact_matrix <- t(contact_data[["matrix"]])
contact_matrix <- contact_matrix / max(Re(eigen(contact_matrix)$values))/demography_vector

# 2. Per-capita formulation
# Normalise the matrix to ensure correct R0 in model
scaling_factor <- 1 / max(Re(eigen(contact_data[["matrix"]])$values))
contact_matrix2 <- contact_data[["matrix.per.capita"]]*scaling_factor
```

So in terms of implications for this episode, the main thing is just to make sure that the notation for the matrix in the model is in terms of contacts susceptibles make with infectives?

Bringing it back to those two key distinctions (which we could tweak to mention, give above discusssion!):
1. **Convert contact matrix into expected number of secondary cases** The R calculation involves an infective meeting susceptibles

2. **Convert contact matrix into contact rates** The epidemic model involves a susceptible meeting infectives

Also tagging @rozeggo and @BlackEdder for info.

Reviewer 2: Do we actually need to distinguish between the two? Given that the eigenvalues will be invariant under transpose (1) could be done in either orientation (as could (2) if the index of the normalising population size is swapped).

Reviewer: There's probably a simpler framing we could use, but think need to explain the two steps (eigenvalue calc, then normalisation by demography over correct matrix dimension), otherwise could lead people to assume the following is correct?

```r
# 1. Current epidemics and finalsize implementation
contact_matrix <- contact_data[["matrix"]] # Edit: no transpose
contact_matrix <- contact_matrix / max(Re(eigen(contact_matrix)$values))/demography_vector
```
Perhaps the below is clearest, because doesn't involve any explicit transposes or normalisation? Just need to explain intuitively what the two matrices represent?
```r
# 2. Per-capita formulation
# Normalise the matrix to ensure correct R0 in model
scaling_factor <- 1 / max(Re(eigen(contact_data[["matrix"]])$values))
contact_matrix2 <- contact_data[["matrix.per.capita"]]*scaling_factor
```

$$
\begin{aligned}
\frac{dS_i}{dt} & = - \beta S_i \sum_j C_{i,j} I_j/N_j \\
\frac{dI_i}{dt} &= \beta S_i\sum_j C_{i,j} I_j/N_j - \gamma I_i \\
\frac{dR_i}{dt} &=\gamma I_i \\
\end{aligned}
$$

###  Normalising the contact matrix to ensure the correct value of $R_0$

When simulating an epidemic, we often want to ensure that the average number of secondary cases generated by a typical infectious individual (i.e. $R_0$) is consistent with known values for the pathogen we're analysing. In the above model, we scale the contact matrix by the $\beta$ to convert the raw interaction data into a transmission rate. But how do we define the value of $\beta$ to ensure a certain value of $R_0$?

Rather than just using the raw number of contacts, we can instead normalise the contact matrix to make it easier to work in terms of $R_0$. In particular, we normalise the matrix by scaling it so that if we were to calculate the average number of secondary cases based on this normalised matrix, the result would be 1 (in mathematical terms, we are scaling the matrix so the largest eigenvalue is 1). This transformation scales the entries but preserves their relative values.

> In the case of the above model, we want to define $\beta  C_{i,j}$ so that the model has a specified valued of $R_0$. The entry of the contact matrix $C[i,j]$ represents the contacts between populations $i$ and $j$, which is equivalent to `contact_data$matrix[j,i]` so the first step is to transpose the contact data matrix (`contact_data$matrix`) so the row/column entries are now in the form $C[i,j]$. Then we normalise the matrix $C$ so the maximum eigenvalue is one and call this matrix $C_normalised$. Because the rate of recovery is $\gamma$, individuals will be infectious on average for $1/\gamma$ days. So $\beta$ is calculated from the scaling factor and the value of $\gamma$  (i.e. mathematically we have the dominant eigenvalue of the matrix $R_0 \times C_{normalised}$ is $\beta / \gamma$).

Reviewer: Suggestion for final version of this paragraph (wrapping up above discussions). I've edited so $C[i,j]$ =`contact_data$matrix[j,i]` as this is how it's defined in line 165 above (and for a training episode, makes it easier to follow the logic).
```
In the case of the above model, we want to define $\beta  C_{i,j}$ so that the model has a specified valued of $R_0$. If the entry of the contact matrix $C[i,j]$ represents the contacts of population $i$ with $j$, it is equivalent to `contact_data$matrix[i,j]`, and the maximum eigenvalue of this matrix represents the typical magnitude of contacts, not typical magnitude of transmission. We must therefore normalise the matrix $C$ so the maximum eigenvalue is one; we call this matrix $C_normalised$. Because the rate of recovery is $\gamma$, individuals will be infectious on average for $1/\gamma$ days. So $\beta$ is calculated from the scaling factor and the value of $\gamma$  (i.e. mathematically we have the dominant eigenvalue of the matrix $R_0 \times C_{normalised}$ is $\beta / \gamma$).
```
